drivers:modem: introducing config for cereg/creg for gsm

mtahirbutt opened the pull request:
As some tests with our 4G modems shown that AT+CREG? does not work as expected with the modems. Therefore, the modem does not get internet connection as the modem expects AT+CREG=0,5 for successful PPP connection but this is never accomplished as some modems do not return AT+CREG=0,5 on successful connection to an operator rather AT+CEREG? is the right AT command to use for such modems. Therefore, a new CONFIG is added to use AT+CEREG? for such modems.
The user can select AT+CREG? or AT+CEREG? according to the type the modem supports.
This PR fixes #63917

mtahirbutt commented:
> Please fix code compliance r

> Please fix code compliance related issues.

thanks sir for the comment, will address tomorrow. thanks for your patience.

emil-lindq commented:
Just a comment, since this is a this or that, maybe this Kconfig symbol should be a choice symbol? In that case please remember to name the choice so a default choice can be selected in some overriding Kconfig.

ycsin reviewed (commented):
would it be possible to use the `COND_CODE1`? I just thought that it might look cleaner

ycsin reviewed (changes requested):
> In that case please remember to name the choice so a default choice can be selected in some overriding Kconfig.

+1

mtahirbutt commented:
> > In that case please remember to name the choice so a default choice can be selected in some overriding Kconfig.
> 
> +
thanks for the remarks, actually different manufacterers have shown different results although all of them are 4G, still I can think about it and work on it if that seems more intuitive to use in modem_gsm_type choice. thanks

mtahirbutt committed:
drivers: modem: introducing config for cereg/creg for gsm

This contribution addresses the support for various types of modems in
gsm driver. As some 4G modems have failed to return correct output
from AT+CREG?, so AT+CEREG? is the right AT command in such situation.
This commit provides the possibility for user to select one type of
AT command. This PR fixes #63917

Signed-off-by: Tahir Akram <mtahirbutt@hotmail.com>

mtahirbutt commented:
> > > In that case please remember to name the choice so a default choice can be selected in some overriding Kconfig.
> 
> thanks for the remarks, actually different manufacterers have shown different results although all of them are 4G, still I can think about it and work on it if that seems more intuitive to use in modem_gsm_type choice. thanks

Please review the changes. I have introduced a choice now. I have also worked with main branch, also thinking of creating a new PR with main branch. As Backport error is occuring here as I already created the issue and mention "Fix #1234"
Thanks.

ycsin commented:
> I have also worked with main branch, also thinking of creating a new PR with main branch

I dont think you can push something into a versioned branch like this, you will have to push things into the `main` branch` and do a proper backport from there

ycsin commented on drivers/modem/Kconfig.gsm:
```diff
@@ -36,6 +36,24 @@ config MODEM_GSM_QUECTEL
+	  Use particular type of AT command for registration status.
+
+config MODEM_GSM_SELECT_CREG
+	bool "CREG command"
```
I think this can be a string type with `"CREG"` as the default value

mtahirbutt replied:
Thanks. I tried but zephyr is not allowing me to use string as a type of choice. i do not find any example either. is there any?

ycsin replied:
oops, didn't know that, how about 
```kconfig
config MODEM_GSM_STATUS_COMMAND
  string "Status command"
  default "CREG" if MODEM_GSM_SELECT_CREG
  default "CEREG" if MODEM_GSM_SELECT_CREG
```
?

I found some similar things here: `zephyrproject/modules/hal/espressif/components/esptool_py/Kconfig.projbuild`

In any case, you should really create another PR against `main`

ycsin commented on drivers/modem/Kconfig.gsm:
```diff
@@ -36,6 +36,24 @@ config MODEM_GSM_QUECTEL
+	  Use this if the modem works only with AT+CREG?.
+
+config MODEM_GSM_SELECT_CEREG
+	bool "CEREG command"
```
and this, with `"CEREG"` as default value

ycsin commented on drivers/modem/Kconfig.gsm:
```diff
@@ -36,6 +36,24 @@ config MODEM_GSM_QUECTEL
 
 endchoice
 
+choice MODEM_GSM_STATUS_COMMAND_TYPE
```
Then this can just be 
```suggestion
choice MODEM_GSM_STATUS_COMMAND
```

ycsin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -523,9 +523,12 @@ MODEM_CMD_DEFINE(on_cmd_atcmdinfo_attached)
-
 static const struct modem_cmd check_net_reg_cmd =
+#if defined(CONFIG_MODEM_GSM_SELECT_CEREG)
+	MODEM_CMD("+CEREG: ", on_cmd_net_reg_sts, 2U, ",");
```
then you  can do this in one liner by concatenating:
`"+" "CONFIG_MODEM_GSM_STATUS_COMMAND"`

ycsin commented on drivers/modem/gsm_ppp.c:
```diff
@@ -764,12 +767,21 @@ static void gsm_finalize_connection(struct k_work *work)
+		ret = modem_cmd_send_nolock(&gsm->context.iface,
+					&gsm->context.cmd_handler,
+					&check_net_reg_cmd, 1,
+					"AT+CEREG?",
```
and:
`"AT+" "CONFIG_MODEM_GSM_STATUS_COMMAND" "?"`

ycsin reviewed (commented):
please create another PR against `main`

mtahirbutt commented:
> > I have also worked with main branch, also thinking of creating a new PR with main branch
> 
> I dont think you can push something into a released branch like this, you will have to push things into the `main` branch` and do a proper backport from there

Please check the new PR #64025. All requested changes addressed. I will close this PR after a while. thanks.

rerickson1 commented:
Moved to a new PR.
